feat(SidebarItem): implement menus for collapsed sidebars

augustobmoura opened the pull request:
[FX-2683](https://toptal-core.atlassian.net/browse/FX-2683)

### Description

When a sidebar is collapsed submenus should be presented as a Dropdown

### How to test

- Check the examples in the storybook for `PageSidebar`, after collapsing a sidebar submenus that were accordions will now open it's submenus as dropdowns instead

### Screenshots

| Before.                                 | After.                                  |
| --------------------------------------- | --------------------------------------- |
| ![image](https://user-images.githubusercontent.com/5845160/166850955-b6992407-ee48-4ebc-a92d-04c05f0d00d8.png) | ![image](https://user-images.githubusercontent.com/5845160/166850707-14fb3673-a05d-45e6-a54a-ed09b4fab251.png) |

### Development checks

- [x] Add changeset according to [guidelines](https://github.com/toptal/picasso/blob/master/docs/contribution/changeset-guidelines.md) (if needed)
- [x] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [x] Annotate all `props` in component with documentation
- [x] Create `examples` for component
- [x] Ensure that deployed demo has expected results and good examples
- [x] Ensure that tests pass by running `yarn test`
- [x] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [x] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).
- [x] Self reviewed
- [x] Covered with tests

**Breaking change**

- [x] codemod is created and showcased in the changeset
- [x] test alpha package of Picasso in StaffPortal

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

augustobmoura commented:
@toptal-bot run all

augustobmoura commented on packages/picasso/src/SidebarItem/SidebarItem.tsx:
```diff
@@ -65,72 +61,47 @@ const useStyles = makeStyles<Theme>(styles, {
   name: 'PicassoSidebarItem'
 })
 
-export const SidebarItem: OverridableComponent<Props> = memo(
```
I split this component into 4 new ones:
- The `Header` that is basically common for all of them, that has the logic of SidebarItemContent (becoming an icon when compact, tooltip and badges)
- A `Basic` version that shows submenus (if any) as a list without any collapsing
- An `Accordion` version that shows submenus as an accordion. It's used when the item has submenus and is `collapsible`
- An `Compact` version that shows submenus as a dropdown. It's used when the item has a submenu and the sidebar is collapsed on the side

augustobmoura replied:
I split this component into 4 new ones:
- The `Header` that is basically common for all of them, that has the logic of SidebarItemContent (becoming an icon when compact, tooltip and badges)
- A `Basic` version that shows submenus (if any) as a list without any collapsing
- An `Accordion` version that shows submenus as an accordion. It's used when the item has submenus and is `collapsible`
- An `Compact` version that shows submenus as a dropdown. It's used when the item has a submenu and the sidebar is collapsed on the side

augustobmoura commented on packages/picasso/src/SidebarItem/SidebarItem.tsx:
```diff
@@ -152,51 +123,131 @@ export const SidebarItem: OverridableComponent<Props> = memo(
+
+    let SidebarItemImpl = BasicSidebarItem
+
+    if (hasMenu) {
```
This is the main logic on switching between the 3 implementations: `Basic`, `Accordion` and `Compact`

denieler replied:
it looks a bit complex, maybe move to a separate component?

```
const SidebarItemWithMenu = (props: Props) => {
  const { compact } = props
  
  return compact
    ? CompactSidebarItem
    : SidebarItemAccordion
}
```
and here will be
```
...
const Component = hasMenu
      ? SidebarItemWithMenu
      : BasicSidebarItem

return <Component {...props} ref={ref} />
```

denieler replied:
maybe like this?

```
    const hasMenu = menu != null

    if (hasMenu && compact) {
      return <CompactSidebarItem {...props} ref={ref} />
    } else if (hasMenu && collapsible) {
      return <SidebarItemAccordion {...props} ref={ref} />
    }

    return <BasicSidebarItem {...props} ref={ref} />
```

or 

```
    const hasMenu = menu != null

    let Component = BasicSidebarItem

    if (hasMenu && compact) {
      Component = CompactSidebarItem
    } else if (hasMenu && collapsible) {
      Component = SidebarItemAccordion
    }

    return <Component {...props} ref={ref} />
```

augustobmoura commented on packages/picasso/src/SidebarItem/SidebarItem.tsx:
```diff
@@ -152,51 +123,131 @@ export const SidebarItem: OverridableComponent<Props> = memo(
+
+    let SidebarItemImpl = BasicSidebarItem
+
+    if (hasMenu) {
```
This is the main logic on switching between the 3 implementations: `Basic`, `Accordion` and `Compact`

denieler reviewed (commented):
`packages/picasso/src/SidebarItem/SidebarItemContext.tsx` file seems to be empty 🤷‍♂️

denieler commented on packages/picasso/src/SidebarItem/SidebarItem.tsx:
```diff
@@ -152,51 +123,131 @@ export const SidebarItem: OverridableComponent<Props> = memo(
+
+    const handleAccordionChange = useCallback(
+      (event: ChangeEvent<{}>, isAccordionExpanded: boolean) => {
+        event.stopPropagation()
```
why do we need to do stopPropagation here?

augustobmoura replied:
It was there before, I just moved it. It looks like it was added on #954. Most probably has an explanation, but I can't tell it 😂

denieler commented on packages/picasso/src/SidebarItem/test.tsx:
```diff
@@ -182,3 +187,23 @@ describe('SidebarItem', () => {
+  describe('when a provider available', () => {
+    it('returns `isSubMenu` as true', () => {
+      const { result } = renderHook(() => useSubMenuContext(), {
+        wrapper: SubMenuContextProvider
```
👍

ozgurkececioglu reviewed (approved):
Tested locally. Working fine 👍 

I would also split the components into different files inside `SidebarItem` folder for cleaner `SidebarItem` component.

denieler reviewed (commented):
found one issue:

![bug](https://user-images.githubusercontent.com/2836281/166967729-6a136e39-3e4c-49b5-a185-b969870d8950.gif)

augustobmoura commented:
> I would also split SidebarItem components into multiple files so it is more clean, but other than that it looks good +1

> Tested locally. Working fine +1
> 
> I would also split the components into different files inside `SidebarItem` folder for cleaner `SidebarItem` component.

Yeah I will split them, I thought about it but because of them sharing the same props decided on keeping them on the same file. I will move `Props` to a `types.ts` file probably

dmaklygin commented on packages/picasso/src/SidebarItem/SidebarItem.tsx:
```diff
@@ -65,72 +61,47 @@ const useStyles = makeStyles<Theme>(styles, {
-      testIds, // eslint-disable-line @typescript-eslint/no-unused-vars
+      onClick,
+      collapsible,
+      // these props are being destructured only for the purpose of excluding them from `...rest`
+      /* eslint-disable @typescript-eslint/no-unused-vars */
+      badge,
+      testIds,
+      icon,
+      isExpanded,
+      expand,
+      index,
```
wdyt it we use three different types of Props, namely: for BasicSidebarItem, for CompactSidebarItem and for BasicSidebarItem, and then the final type Props will be a concatenation of these three types?

augustobmoura replied:
Yeah, I will try to do it

dmaklygin commented on packages/picasso/src/SidebarItem/SidebarItem.tsx:
```diff
@@ -152,51 +123,131 @@ export const SidebarItem: OverridableComponent<Props> = memo(
-      )
-    }
+const BasicSidebarItem = forwardRef<HTMLElement, Props>(
+  function BasicSidebarItem(props: Props, ref) {
```
because, it's weird to expect here `expand` prop, because it's not used here. However, it's defined in `Props` deifnition

OndrejTuma commented on packages/picasso/src/PageSidebar/PageSidebar.tsx:
```diff
@@ -91,8 +97,16 @@ export const PageSidebar = forwardRef<HTMLDivElement, Props>(function Sidebar(
+    }
+  }, [isCollapsed])
+
+  useEffect(() => {
```
You sure its okay to remove layoutEffect from here?

augustobmoura replied:
We talked about it `here`. It should be ok to use `useEffect` instead of `useLayoutEffect`, as is just setting a flag, and nothing complex. I will take a better look

OndrejTuma reviewed (approved):
Tested and works fine 👍  code looks good 🎉
